File: DesignPattenStudy/ProxyPattern/Java/ProxyRider.java
package ProxyPattern.Java;

public class ProxyRider implements Rider{

    private FullTimeRider fullTimeRider = null;
    private String name;

    public ProxyRider(String name) {

        this.name = name;
    }

    @Override
    public String delivery() {

        if(fullTimeRider == null) {

            fullTimeRider = new FullTimeRider(name);
        }

        return fullTimeRider.delivery();
    }

    @Override
    public String getPersonalInfo(Integer level) {

        return name;
    }
}

File: DesignPattenStudy/ProxyPattern/Kotlin/ProxyRider.kt
package ProxyPattern.Kotlin

class ProxyRider(var name: String) : Rider {

    private var fullTimeRider: FullTimeRider? = null

    override fun delivery(): String =
            (fullTimeRider ?: FullTimeRider(name)).delivery()

    override fun getPersonInfo(level: Int): String? = name
}